### PR TITLE
14 create and store a hedera subscription

### LINKED ISSUE #1: EI: Parse TX memo field
- Parse TX memo field of TXs received from Mirror Node Polling. 
- Extract TaskId and TopicId from memo field. 
- Map this TaskId to some JobSpec id on the Chainlink Node.

### LINKED ISSUE #2: EI: Initiate (trigger) a specific job
- upon successful parsing of a TX memo and mapping to a subscription
- Initiate (trigger) a specific job in the Chainlink Node 
- Pass the required parameter - topic id for the response that was extracted from the TX memo field.

### LINKED ISSUE #9: EI: Verify payment.
Implement logic to verify if a correct payment was made as part of the Hedera Transaction.
This should be done on EI side upon receiving a transaction from the Mirror Node, before extracting JobID and TopicID from the TX memo field.

### LINKED ISSUE #10: EI: Mirror Node Polling
- Implement polling of Hedera Mirror Node e.g. every 5 seconds for new transactions to specific address
- If we use this address for another subscription  there is no need to start polling for TXs to the same address
- Use timestamps to query for transactions after a specific time - the previous request made.
- Map received TX jobID to active subscription JobID

### LINKED ISSUE #14: EI: Create and store a hedera subscription.
Implement logic to create and store a hedera subscription.
Changes need to be made to:
- hedera.go - Implement meaningful Test() function - this is used by service.go when creating a new subscription.
- add timestamp field to hedera_subscription type in database.
- create logic to receive new requests from the Chainlink Node upon new job definition as described here: https://www.notion.so/limechain/Chainlink-JobId-EI-subscription-Mirror-Node-Polling-design-1d21bf56551d4b46aff40c8dae686c24

### PR DESCRIPTION
**Pull request by iamrusko**

### CODE REVIEW

File: blockchain/hedera.go [hunk 1]
```diff
@@ -147,6 +194,75 @@ func (c Client) GetAccountCreditTransactionsAfterTimestamp(accountID string, fro
...
+					logger.Errorf(err.Error())
+					return
+				}
+				lastTransactionTimestamp = transactionTimestamp
+			}
+			for _, transaction := range response.Transactions {
+				// This request is targeting a specific jobID
+				decodedMemo, err := DecodeMemo(transaction.MemoBase64)
+				if err != nil {
+					logger.Error("Failed decoding base64 NEAROracleRequestArgs.RequestSpec:", err)
```
Comment (failfmi) [inline blockchain/hedera.go]: continue?

File: blockchain/hedera.go [hunk 2]
```diff
@@ -147,6 +194,75 @@ func (c Client) GetAccountCreditTransactionsAfterTimestamp(accountID string, fro
...
+				transactionTimestamp, err := FromString(response.Transactions[numberOfTransactions-1].ConsensusTimestamp)
+				if err != nil {
+					logger.Errorf(err.Error())
+					return
+				}
+				lastTransactionTimestamp = transactionTimestamp
+			}
+			for _, transaction := range response.Transactions {
+				// This request is targeting a specific jobID
+				decodedMemo, err := DecodeMemo(transaction.MemoBase64)
```
Comment (failfmi) [inline blockchain/hedera.go:280]: Before you decode it, you can check if there is anything - `len(transaction.MemoBase64) == 0` -> skip

File: blockchain/hedera.go [hunk 3]
```diff
@@ -57,21 +72,35 @@ func (hSubscr hederaSubscriber) SubscribeToEvents(channel chan<- subscriber.Even
 	//todo map somehow task id from memo to a different jobspec? is that possible? what is jobID in hederaSubscription?
 	//todo check if payment was okay
 	//todo create tests - hedera_test.go & also elsewhere where we changed the code + db migration
+	//todo remove last transaction timestamp from hedera_subscription database object if we're not going to save it
```
Comment (stoqnkpL) [inline blockchain/hedera.go]: I think we've already implemented these //TODOs and can delete them, right? Also on line 115

File: blockchain/hedera.go [hunk 4]
```diff
@@ -147,6 +194,75 @@ func (c Client) GetAccountCreditTransactionsAfterTimestamp(accountID string, fro
...
 }
 
+func (c Client) GetAccountByAccountId(accountID string) (*Response, error) {
+	accountQuery := fmt.Sprintf("?account.id=%s", accountID)
+	return c.getAccountByQuery(accountQuery)
+}
+
+func DecodeTransactionMemo(transactionMemo string) ([]byte, error) {
+	return base64.StdEncoding.DecodeString(transactionMemo)
+}
```
Comment (stoqnkpL) [inline blockchain/hedera.go]: this in never called, we can remove it

File: blockchain/hedera.go [hunk 5]
```diff
@@ -181,6 +297,39 @@ func ToHumanReadable(timestampNanos int64) string {
 	return parsed.Format(time.RFC3339Nano)
 }
```
Comment (stoqnkpL) [inline blockchain/hedera.go]: this in never called, we can remove it

File: blockchain/hedera.go (hunk 3 above)
Comment (iamrusko) [reply blockchain/hedera.go]: removed

File: blockchain/common.go [hunk 6]
```diff
@@ -54,6 +54,7 @@ type Params struct {
 	Addresses   []string `json:"addresses"`
 	Topics      []string `json:"topics"`
 	AccountIds  []string `json:"accountIds"`
+	AccountId   string 	 `json:"accountId"`
```
Comment (failfmi) [inline blockchain/common.go]: Can we format this

File: blockchain/hedera.go [hunk 7]
```diff
@@ -15,63 +17,129 @@ import (
...
-	accountIds   []string
+	accountId   string
 	monitorResp *http.Response
 	isDone      bool
 	jobid       string
 }
 
-func (hSubscr hederaSubscriber) SubscribeToEvents(channel chan<- subscriber.Event, _ store.RuntimeConfig) (subscriber.ISubscription, error) {
+type HederaConfig struct {
+	LinkTopicId string `mapstructure:"LINK_TOPIC_ID"`
```
Comment (failfmi) [inline blockchain/hedera.go]: I think this should be `TOKEN_ID`

File: blockchain/hedera.go [hunk 8]
```diff
@@ -147,6 +234,79 @@ func (c Client) GetAccountCreditTransactionsAfterTimestamp(accountID string, fro
...
+				extractedTopicId := hederaTransactionInfo[0]
+				extractedJobId := hederaTransactionInfo[1]
+
+				if !containsJobId(hederaSubscribersMap[accoutId], extractedJobId) {
+					continue
+				} else {
+					hederaSubs := hederaSubscribersMap[accoutId]
+					for _, hs := range hederaSubs {
+						if hs.jobid == extractedJobId {
+							eventInfo := fmt.Sprintf("{\"hederaTopicId\": \"%s\"}", extractedTopicId)
```
Comment (failfmi) [inline blockchain/hedera.go]: Can we switch this to a struct, which will later be marshalled?

File: blockchain/hedera.go [hunk 9]
```diff
@@ -15,63 +17,129 @@ import (
...
+					return errors.New(errorMessage)
+				}
+			} else {
+				errorMessage := fmt.Sprintf("Account with ID: %s is not assigned to LINK token with ID: %s", hSubscr.AccountId, tokenId)
+				return errors.New(errorMessage)
+			}
+		} else {
+			return errors.New("Please check that you have recorded the Hedera's AccountId correctly")
+		}
+	}
```
Comment (failfmi) [inline blockchain/hedera.go]: Can we refactor the logic to not be that nested?

File: blockchain/hedera.go [hunk 10]
```diff
@@ -175,10 +335,56 @@ func String(timestamp int64) string {
...
+	decodedMemo, err := base64.StdEncoding.DecodeString(base64Memo)
+	if err != nil {
+		return "", err
+	}
+
+	return string(decodedMemo), nil
+}
+
+func checkForValidTokenTransfer(tokenTransfers []Transfer, accountId string) bool {
+
```
Comment (failfmi) [inline blockchain/hedera.go]: nit: remove empty line

File: store/database.go [hunk 11]
```diff
@@ -375,6 +374,6 @@ type AgoricSubscription struct {
 
 type HederaSubscription struct {
 	gorm.Model
-	SubscriptionId uint
-	AccountIds     SQLStringArray
+	SubscriptionId			 uint
+	AccountId				 string
```
Comment (failfmi) [inline store/database.go]: nit: format code

File: store/migrations/migration1614545435/migrate.go [hunk 12]
```diff
@@ -17,8 +17,8 @@ import (
 
 type HederaSubscription struct {
 	gorm.Model
-	SubscriptionId uint
-	AccountIds      string
+	SubscriptionId 			 uint
+	AccountId      			 string
```
Comment (failfmi) [inline store/migrations/migration1614545435/migrate.go]: nit: format code

File: blockchain/hedera.go [hunk 13]
```diff
@@ -147,6 +234,79 @@ func (c Client) GetAccountCreditTransactionsAfterTimestamp(accountID string, fro
 	return c.getTransactionsByQuery(transactionsDownloadQuery)
 }
 
+func (c Client) GetAccountByAccountId(accountID string) (*Response, error) {
+	accountQuery := fmt.Sprintf("?account.id=%s", accountID)
+	return c.getAccountByQuery(accountQuery)
+}
+
+// WaitForTransaction Polls the transaction at intervals.
+func (c Client) WaitForTransaction(accoutId string) {
```
Comment (failfmi) [inline blockchain/hedera.go]: nit: `accoutId` -> `accountId`

File: blockchain/hedera.go [hunk 14]
```diff
@@ -175,10 +335,56 @@ func String(timestamp int64) string {
...
+
+	return string(decodedMemo), nil
+}
+
+func checkForValidTokenTransfer(tokenTransfers []Transfer, accountId string) bool {
+
+	if tokenId == "" {
+		logger.Error("LINK Token ID is missing! Please set LINK Token ID to .env configuration file.")
+		return false
+	}
```
Comment (failfmi) [inline blockchain/hedera.go]: Is this check necessary?

File: blockchain/hedera.go (hunk 8 above)
Comment (iamrusko) [reply blockchain/hedera.go]: done

File: blockchain/hedera.go (hunk 7 above)
Comment (iamrusko) [reply blockchain/hedera.go]: done

File: blockchain/common.go (hunk 6 above)
Comment (iamrusko) [reply blockchain/common.go]: done

File: blockchain/hedera.go (hunk 9 above)
Comment (iamrusko) [reply blockchain/hedera.go]: done

File: blockchain/hedera.go (hunk 10 above)
Comment (iamrusko) [reply blockchain/hedera.go]: done

File: store/database.go (hunk 11 above)
Comment (iamrusko) [reply store/database.go]: done

File: store/migrations/migration1614545435/migrate.go (hunk 12 above)
Comment (iamrusko) [reply store/migrations/migration1614545435/migrate.go]: done

File: blockchain/hedera.go (hunk 13 above)
Comment (iamrusko) [reply blockchain/hedera.go]: done

File: blockchain/hedera.go (hunk 14 above)
Comment (iamrusko) [reply blockchain/hedera.go]: removed

**Review by failfmi — APPROVED** [review]

**Pull request merged**